src/rrd_daemon.c

```diff
@@ -2385,6 +2386,103 @@ static int handle_request_create (HANDLER_PROTO) /* {{{ */
   return rc;
 } /* }}} static int handle_request_create  */
 
+static int handle_request_list (HANDLER_PROTO) /* {{{ */
+{
+  char *filename;
+  char fullpath[PATH_MAX];
+  int status;
+  DIR *dir;
+  struct dirent *entry;
+
+  if (config_base_dir == NULL) {
+    return send_response(sock, RESP_ERR, "No base directory defined\n");
+  }
+
+  /* Get pathname */
+  status = buffer_get_field(&buffer, &buffer_size, &filename);
+
+  if (status != 0)
+    return syntax_error(sock,cmd);
+
+  /* get full pathname */
+  snprintf(fullpath, PATH_MAX, "%s%s%s",
+	   config_base_dir, (filename[0] == '/') ? "" : "/", filename);
+
+  /* Prevent moving up the directory tree */
+  if (strstr(fullpath, ".."))
+      send_response(sock, RESP_ERR, "%s\n", rrd_strerror(EACCES));
+
+  /* if filename contains wildcards, then use glob() */
+  if (strchr(fullpath, '*') || strchr(fullpath, '?')) {
+    glob_t buf;
+    unsigned int i;
+
+    if (glob(fullpath, 0, NULL, &buf)) {
+      globfree(&buf);
+      goto out_send_response;
+    }
+
+    for (i = 0; i < buf.gl_pathc; i++) {
+      char *ptr;
+
+      if (!check_file_access(buf.gl_pathv[i], sock))
+	return send_response(sock, RESP_ERR,
+			     "Cannot read: %s\n", buf.gl_pathv[i]);
+
+      ptr = strrchr(buf.gl_pathv[i], '/');
+
+      if (ptr != NULL)
+	add_response_info(sock, "%s\n", ptr + 1);
+    }
+
+    globfree(&buf);
+    goto out_send_response;
+  }
+
+  if (!check_file_access(fullpath, sock))
+    return send_response(sock, RESP_ERR, "Cannot read: %s\n", fullpath);
+
+  /* If filename matches an RRD file, then return it */
+  if (strstr(fullpath, ".rrd")) {
+	  struct stat st;
+	  char *ptr;
+
+	  if (!stat(fullpath, &st) && S_ISREG(st.st_mode)) {
+		  ptr = strrchr(fullpath, '/');
+
+		  if (ptr)
+			  add_response_info(sock, "%s\n", ptr + 1);
+	  }
+
+	  goto out_send_response;
+  }
+
+  dir = opendir(fullpath);
+
+  if (dir == NULL) {
+    return send_response(sock, RESP_ERR,
+			 "Failed to open directory %s : %s\n",
+			 fullpath, strerror(errno));
+  }
+
+  while ((entry = readdir(dir)) != NULL) {
+
+    if ((strcmp(entry->d_name, ".") == 0) ||
+	(strcmp(entry->d_name, "..") == 0)) {
+      continue;
+    }
+
```

Comment:
it will also show non rrd files  ... how about only showing directories and files ending in .rrd